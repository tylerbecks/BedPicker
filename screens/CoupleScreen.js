import React from "react";
import _ from "lodash";
import { ScrollView, StyleSheet, Text, View } from "react-native";
import CoupleGallery from "../components/CoupleGallery";
import SelectButton from "../components/SelectButton";
import SubmitButton from "../components/SubmitButton";
import convertGuestsToSleepers from "../utils/convertGuestsToSleepers";

const styles = StyleSheet.create({
  guestSelectButtonsContainer: {
    flexDirection: "row",
    flexWrap: "wrap",
    justifyContent: "center"
  },
  container: {
    backgroundColor: "#fff",
    flex: 1
  },
  contentContainer: {
    paddingHorizontal: 20,
    paddingTop: 30
  },
  titleText: {
    fontSize: 24,
    fontWeight: "bold",
    textAlign: "center"
  },
  plusSign: {
    fontSize: 32
  }
});

const getCoupleKey = (guest, coupledGuestsIdMap) => {
  let coupleKey;

  _.each(coupledGuestsIdMap, (value, key) => {
    if (value === guest.id || key === guest.id) {
      coupleKey = key;
      return false; // exit loop early
    }
  });

  return coupleKey;
};

const removeCouple = (selectedGuest, coupledGuestsIdMap) => {
  const coupleKey = getCoupleKey(selectedGuest, coupledGuestsIdMap);
  if (!coupleKey) {
    throw Error("coupleKey should exist");
  }

  return {
    coupledGuestsIdMap: _.omit(coupledGuestsIdMap, coupleKey)
  };
};

const addCouple = (selectedGuest, stagedGuest, coupledGuestsIdMap) => {
  return {
    stagedGuest: null,
    coupledGuestsIdMap: {
      ...coupledGuestsIdMap,
      [stagedGuest.id]: selectedGuest.id
    }
  };
};

export default class CoupleScreen extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      coupledGuestsIdMap: {},
      stagedGuest: null
    };

    this.onSubmit = this.onSubmit.bind(this);
  }

  onPressGuest(selectedGuest) {
    this.setState(({ coupledGuestsIdMap, stagedGuest }) => {
      const isSelectedGuestInStaging = stagedGuest === selectedGuest;
      const isSelectedGuestCoupled = getCoupleKey(
        selectedGuest,
        coupledGuestsIdMap
      );

      if (isSelectedGuestInStaging) {
        return { stagedGuest: null };
      }
      if (isSelectedGuestCoupled) {
        return removeCouple(selectedGuest, coupledGuestsIdMap);
      }
      if (stagedGuest) {
        return addCouple(selectedGuest, stagedGuest, coupledGuestsIdMap);
      }

      return { stagedGuest: selectedGuest };
    });
  }

  onSubmit() {
    const { navigation } = this.props;
    const { coupledGuestsIdMap } = this.state;
    const selectedGuests = this.getSelectedGuests();
    const sleepers = convertGuestsToSleepers(
      selectedGuests,
      coupledGuestsIdMap
    );

    navigation.navigate("Assignment", { sleepers });
  }

  getSelectedGuests() {
    const { navigation } = this.props;

    return navigation.getParam("selectedGuests", []);
  }

  getCouples() {
    const { coupledGuestsIdMap } = this.state;
    const selectedGuests = this.getSelectedGuests();

    const coupleIds = _.toPairs(coupledGuestsIdMap);
    return _.map(coupleIds, ids => [
      _.find(selectedGuests, { id: ids[0] }),
      _.find(selectedGuests, { id: ids[1] })
    ]);
    // map over pairs and get couples
  }

  render() {
    const { stagedGuest, coupledGuestsIdMap } = this.state;
    const selectedGuests = this.getSelectedGuests();
    const couples = this.getCouples();

    return (
      <View style={styles.container}>
        <ScrollView
          style={styles.container}
          contentContainerStyle={styles.contentContainer}
        >
          <Text style={styles.titleText}>Select couples</Text>

          <CoupleGallery couples={couples} />

          <View style={styles.guestSelectButtonsContainer}>
            {selectedGuests.map(guest => (
              <SelectButton
                key={guest.name}
                text={guest.name}
                photo={guest.photo}
                onPress={() => this.onPressGuest(guest)}
                selected={
                  guest === stagedGuest ||
                  getCoupleKey(guest, coupledGuestsIdMap)
                }
              />
            ))}
          </View>
          <SubmitButton
            onPress={this.onSubmit}
            disabled={false}
            text="Submit"
          />
        </ScrollView>
      </View>
    );
  }
}
