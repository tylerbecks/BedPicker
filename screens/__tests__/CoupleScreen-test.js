import 'react-native';
import React from 'react';
import CoupleScreen from '../CoupleScreen';
import renderer from 'react-test-renderer';

const navigationStub = {
  getParam() {
    return [];
  }
};

it('renders', () => {
  expect(
    renderer.create(<CoupleScreen navigation={navigationStub} />).toJSON()
  ).toMatchSnapshot();
});
