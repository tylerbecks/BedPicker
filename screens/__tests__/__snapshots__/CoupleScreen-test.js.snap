// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#fff",
      "flex": 1,
    }
  }
>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "paddingHorizontal": 20,
        "paddingTop": 30,
      }
    }
    style={
      Object {
        "backgroundColor": "#fff",
        "flex": 1,
      }
    }
  >
    <View>
      <Text
        style={
          Object {
            "fontSize": 24,
            "fontWeight": "bold",
            "textAlign": "center",
          }
        }
      >
        Select couples
      </Text>
      <View
        style={
          Object {
            "flexDirection": "row",
            "flexWrap": "wrap",
            "marginBottom": 20,
          }
        }
      />
      <View
        style={
          Object {
            "flexDirection": "row",
            "flexWrap": "wrap",
            "justifyContent": "center",
          }
        }
      />
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#21ba45",
            "borderRadius": 100,
            "height": 80,
            "justifyContent": "center",
            "margin": 10,
            "opacity": 1,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 20,
              "fontWeight": "bold",
            }
          }
        >
          Submit
        </Text>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
